In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Load Titanic bằng seaborn theo yêu cầu BẮT BUỘC
df_titanic = sns.load_dataset('titanic')

# Chạy shape, head, tail, info, describe
print("1. Shape:", df_titanic.shape)
print("\n2. Head:")
display(df_titanic.head())
print("\n3. Tail:")
display(df_titanic.tail())
print("\n4. Info:")
display(df_titanic.info())
print("\n5. Describe:")
display(df_titanic.describe())

1. Shape: (891, 15)

2. Head:


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True



3. Tail:


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
886,0,2,male,27.0,0,0,13.00,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.00,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.45,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.00,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.75,Q,Third,man,True,NaN,Queenstown,no,True



4. Info:
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    str     
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    str     
 8   class        891 non-null    category
 9   who          891 non-null    str     
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    str     
 13  alive        891 non-null    str     
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), str(5)
memory usage: 80.7 KB


None


5. Describe:


,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


**Giải thích cấu trúc dữ liệu cơ bản:**
* **Về Shape:** Tập dữ liệu có tổng cộng 891 dòng và 15 cột. Số lượng dòng đủ lớn để đưa ra các kết luận thống kê có ý nghĩa.
* **Về Head:** Dữ liệu 5 dòng đầu cho thấy biến mục tiêu là `survived` (0 và 1). Có sự pha trộn giữa dữ liệu số (tuổi, giá vé) và dữ liệu chuỗi (giới tính, thành phố).
* **Về Tail:** 5 dòng cuối cùng cũng cho thấy sự hiện diện của rất nhiều giá trị NaN ở cột `deck` (mã cabin). Cấu trúc dữ liệu nhất quán từ đầu đến cuối file.
* **Về Info:** Có 3 cột đang bị thiếu dữ liệu là `age`, `deck` và `embarked_town`. Các cột còn lại có kiểu dữ liệu (int, float, object) hoàn toàn hợp lý.
* **Về Describe:** Trung bình chỉ có 38.38% hành khách sống sót. Độ tuổi trải dài từ trẻ sơ sinh (0.42) đến 80 tuổi, với giá vé có độ lệch rất lớn (từ 0 đến 512 đô la).

In [2]:
# Kiểm tra missing values
missing_count = df_titanic.isna().sum()
missing_percent = (missing_count / len(df_titanic)) * 100
missing_df = pd.DataFrame({'Số lượng thiếu': missing_count, 'Phần trăm thiếu (%)': missing_percent})
missing_df[missing_df['Số lượng thiếu'] > 0].sort_values(by='Phần trăm thiếu (%)', ascending=False)

,Số lượng thiếu,Phần trăm thiếu (%)
deck,688,77.216611
age,177,19.865320
embarked,2,0.224467
embark_town,2,0.224467


**Nhận xét Missing Values:**
* **Cột `deck` (Cabin):** Đây là cột có vấn đề nghiêm trọng nhất vì bị thiếu tới 77.2% dữ liệu. Lý do là trong tình trạng hỗn loạn khi chìm tàu, thông tin về mã phòng của hành khách không được ghi chép hoặc lưu giữ lại đầy đủ.
* **Cột `age` (Tuổi):** Bị thiếu 19.8% dữ liệu (177 dòng). Đây là một biến quan trọng, vấn đề này có thể được khắc phục bằng cách điền các giá trị độ tuổi trung vị của toàn tàu vào những dòng bị trống.

In [3]:
# Phân tích 3 cột Categorical
print("Phân bố Giới tính (sex):\n", df_titanic['sex'].value_counts(), "\n")
print("Phân bố Hạng vé (pclass):\n", df_titanic['pclass'].value_counts(), "\n")
print("Phân bố Cảng lên tàu (embarked):\n", df_titanic['embarked'].value_counts())

Phân bố Giới tính (sex):
 sex
male      577
female    314
Name: count, dtype: int64 

Phân bố Hạng vé (pclass):
 pclass
3    491
1    216
2    184
Name: count, dtype: int64 

Phân bố Cảng lên tàu (embarked):
 embarked
S    644
C    168
Q     77
Name: count, dtype: int64


**Giải thích phân bố Categorical:** * **sex:** Lượng khách nam giới (577 người) đông gần gấp đôi nữ giới (314 người), phản ánh bối cảnh lao động và di cư thời kỳ đó.
* **pclass:** Phần lớn hành khách là người lao động nghèo mua vé Hạng 3 (491 người), áp đảo hoàn toàn so với tầng lớp trung lưu (Hạng 2) và thượng lưu (Hạng 1).
* **embarked:** Cảng Southampton (S) là điểm khởi hành chính của con tàu nên đón lượng khách lên tàu đông nhất (644 người).

In [4]:
# Phân tích 3 cột Numeric
numeric_cols = ['age', 'fare', 'sibsp']
display(pd.DataFrame({
    'Mean': df_titanic[numeric_cols].mean(),
    'Median': df_titanic[numeric_cols].median(),
    'Min': df_titanic[numeric_cols].min(),
    'Max': df_titanic[numeric_cols].max()
}))

,Mean,Median,Min,Max
age,29.699118,28.0000,0.42,80.0000
fare,32.204208,14.4542,0.00,512.3292
sibsp,0.523008,0.0000,0.00,8.0000


**Nhận xét điểm bất thường Numeric:** * **Cột `fare`:** Có điểm cực kỳ bất thường khi giá vé Min = 0.0 (đi miễn phí) và Max = 512.3 đô la, khiến giá trị trung bình (Mean = 32.2) bị kéo lên cao gấp đôi so với trung vị (Median = 14.45).
* **Cột `sibsp`:** Điểm bất thường là có những hành khách đi cùng một đại gia đình lên tới 8 người anh chị em/vợ chồng (Max = 8), trong khi đa số mọi người (Median = 0) đều đi một mình.

In [5]:
# Tính tỉ lệ sống sót toàn tàu
ti_le_song_sot = df_titanic['survived'].mean() * 100
print(f"Tỉ lệ sống sót toàn tàu: {ti_le_song_sot:.2f}%")

Tỉ lệ sống sót toàn tàu: 38.38%


**Ý nghĩa con số:** Chỉ có khoảng 38.38% trên tổng số 891 hành khách trong tập dữ liệu này sống sót. Con số này cho thấy mức độ tàn khốc của thảm họa Titanic, hơn 60% số người trên tàu đã không thể qua khỏi do thiếu thuyền cứu sinh.

In [6]:
# Groupby tỉ lệ sống sót theo giới tính và hạng vé
print("Tỉ lệ sống sót theo Giới tính (sex):")
display(df_titanic.groupby('sex')['survived'].mean() * 100)

print("\nTỉ lệ sống sót theo Hạng vé (pclass):")
display(df_titanic.groupby('pclass')['survived'].mean() * 100)

Tỉ lệ sống sót theo Giới tính (sex):


sex
female    74.203822
male      18.890815
Name: survived, dtype: float64


Tỉ lệ sống sót theo Hạng vé (pclass):


pclass
1    62.962963
2    47.282609
3    24.236253
Name: survived, dtype: float64

**Nhận xét so sánh:** Sự phân hóa sống sót là vô cùng rõ rệt. Về giới tính, nữ giới có tỉ lệ sống (74.2%) cao hơn rất nhiều so với nam giới (18.8%). Về giai cấp, hành khách khoang hạng 1 (giới giàu có) có tỉ lệ sống sót cao nhất (62.9%), trong khi người nghèo ở khoang hạng 3 lại thiệt thòi nhất (chỉ 24.2%).

In [7]:
# Trả lời 3 câu hỏi từ Phần A (Câu 1, 2, 3)
# - Câu 1 và 2 đã được chứng minh qua lệnh groupby bên trên.
# - Dưới đây là code trả lời Câu 3: Trẻ em (<16 tuổi) có được ưu tiên không?
df_titanic['is_child'] = df_titanic['age'] < 16
print("Tỉ lệ sống sót của Trẻ em (True) so với Người lớn (False):")
display(df_titanic.groupby('is_child')['survived'].mean() * 100)

Tỉ lệ sống sót của Trẻ em (True) so với Người lớn (False):


is_child
False    36.262376
True     59.036145
Name: survived, dtype: float64

**Đối chiếu dự đoán vs thực tế:**
* **Câu 1 (Giới tính):** [ĐÚNG] Thực tế tỉ lệ sống của nữ vượt xa nam. 
* **Câu 2 (Hạng vé):** [ĐÚNG] Hành khách Hạng 1 được ưu tiên sống sót cao nhất.
* **Câu 3 (Trẻ em):** [ĐÚNG] Trẻ em (True) có tỉ lệ sống sót lên tới 59.0%, cao hơn mức 38.1% của nhóm người lớn.
* **Nhận xét kết luận:** Những trực giác ban đầu hoàn toàn trùng khớp với thực tế dữ liệu. Các nguyên tắc về đạo đức ("phụ nữ và trẻ em trước") cũng như sự phân hóa giàu nghèo đã chi phối trực tiếp đến cơ hội sống sót của hành khách trong thảm họa này.

# PHẦN D – CÂU HỎI MỚI (SAU PHÂN TÍCH)

1. Tỉ lệ sống sót của nhóm phụ nữ ở khoang Hạng 3 có cao hơn nhóm nam giới ở khoang Hạng 1 hay không?
2. Những người đi một mình (thuộc tính alone = True) có tỉ lệ sống sót thấp hơn những người đi có gia đình không?
3. Việc bước lên tàu ở các bến cảng khác nhau (embarked) có vô tình tạo ra sự chênh lệch về tỉ lệ sống sót không?
4. Những hành khách có thông tin mã phòng (không bị missing ở cột deck) có phải là nhóm khách VIP có tỉ lệ sống sót cao hơn không?
5. Mức giá vé (fare) trung bình của nhóm những người sống sót chênh lệch bao nhiêu đô la so với nhóm tử vong?

In [8]:
# Thử trả lời 2 trong 5 câu hỏi mới bằng code
# 1. Trả lời Câu mới số 2: Đi một mình (alone)
print("Tỉ lệ sống sót giữa đi một mình (True) và có gia đình (False):")
display(df_titanic.groupby('alone')['survived'].mean() * 100)

# 2. Trả lời Câu mới số 5: Giá vé trung bình
print("\nGiá vé trung bình của nhóm Tử vong (0) và Sống sót (1):")
display(df_titanic.groupby('survived')['fare'].mean())

Tỉ lệ sống sót giữa đi một mình (True) và có gia đình (False):


alone
False    50.564972
True     30.353818
Name: survived, dtype: float64


Giá vé trung bình của nhóm Tử vong (0) và Sống sót (1):


survived
0    22.117887
1    48.395408
Name: fare, dtype: float64

**Nhận xét (Bonus):**
* Nhóm đi cùng người thân (alone = False) có tỉ lệ sống cao hơn (50.5%) so với những người đi một mình (alone = True, chỉ đạt 30.3%).
* Những người sống sót có mức giá vé trung bình (48.3 đô) cao gấp đôi những người tử vong (22.1 đô). Tiền bạc rõ ràng có tác động lớn đến cơ hội sống sót.